pkg/apiserver/registry/system/supportbundle/rest.go

```diff
@@ -0,0 +1,378 @@
+// Copyright 2020 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package supportbundle
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"context"
+	"crypto/sha256"
+	"fmt"
+	"io"
+	"math/rand"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/spf13/afero"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apiserver/pkg/registry/rest"
+	"k8s.io/klog"
+	"k8s.io/utils/exec"
+
+	agentutil "github.com/vmware-tanzu/antrea/pkg/agent/util"
+	systemv1beta1 "github.com/vmware-tanzu/antrea/pkg/apis/system/v1beta1"
+	controllerutil "github.com/vmware-tanzu/antrea/pkg/controller/util"
+	"github.com/vmware-tanzu/antrea/pkg/ovs/ovsctl"
+	"github.com/vmware-tanzu/antrea/pkg/util"
+)
+
+const bundleExpireDuration = time.Hour
+
+var (
+	defaultFS       = afero.NewOsFs()
+	defaultExecutor = exec.New()
+)
+
+// NewStorage creates a support bundle storage. The mode should be either agent
+// or controller. If the mode is agent, the client argument should not be nil.
+func NewStorage(mode string, client ovsctl.OVSCtlClient) Storage {
+	bundle := &supportBundleREST{
+		mode:         mode,
+		ovsCtlClient: client,
+		cache: &supportBundle{
+			SupportBundle: systemv1beta1.SupportBundle{
+				ObjectMeta: metav1.ObjectMeta{Name: mode},
+				Status:     systemv1beta1.SupportBundleStatusNone,
+			},
+		},
+	}
+	return Storage{
+		Mode:          mode,
+		SupportBundle: bundle,
+		Download:      &downloadREST{supportBundle: bundle},
+	}
+}
+
+// Storage contains REST resources for support bundle, includes status query and download.
+type Storage struct {
+	SupportBundle *supportBundleREST
+	Download      *downloadREST
+	Mode          string
+}
+
+var (
+	_ rest.Scoper          = &supportBundleREST{}
+	_ rest.Getter          = &supportBundleREST{}
+	_ rest.Creater         = &supportBundleREST{}
+	_ rest.GracefulDeleter = &supportBundleREST{}
+)
+
+type supportBundle struct {
+	systemv1beta1.SupportBundle
+	filepath string
+}
+
+// supportBundleREST implements REST interfaces for bundle status querying.
+type supportBundleREST struct {
+	bridge       string
+	mode         string
+	statusLocker sync.RWMutex
+	cancelFunc   context.CancelFunc
+	cache        *supportBundle
+	ovsCtlClient ovsctl.OVSCtlClient
+}
+
+// Create triggers a bundle generation progress. It only allows to create resource
```

Comment:
did you mean "progress" or "process"?

Reply:
Thanks, I would like to remove this word.